packages/graphql/src/schema/validation/schema-validation.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type {
+    DocumentNode,
+    DefinitionNode,
+    GraphQLDirective,
+    GraphQLNamedType,
+    GraphQLError,
+} from "graphql";
+import { visit, visitInParallel, specifiedDirectives, GraphQLSchema } from "graphql";
+import { getStaticAuthorizationDefinitions } from "../../graphql/directives/type-dependant-directives/authorization";
+import { authorizationDefinitionsEnricher, authorizationUsageEnricher } from "./enrichers/authorization";
+import { EnricherContext } from "./EnricherContext";
+import type { Enricher } from "./types";
+import { specifiedSDLRules } from "graphql/validation/specifiedRules";
+import type { SDLValidationRule } from "graphql/validation/ValidationContext";
+import { SDLValidationContext } from "graphql/validation/ValidationContext";
+import type { Maybe } from "graphql/jsutils/Maybe";
+
+function getAdditionalDefinitions(): DefinitionNode[] {
+    return getStaticAuthorizationDefinitions();
+}
+function enrichDocument(
+    enrichers: Enricher[],
+    additionalDefinitions: DefinitionNode[],
+    document: DocumentNode
+): DocumentNode {
+    return {
+        ...document,
+        definitions: enrichers
+            .reduce(
+                (definitions, enricher) => definitions.reduce(enricher, [] as DefinitionNode[]),
+                document.definitions
+            )
+            .concat(...additionalDefinitions),
+    };
+}
+
+function makeValidationDocument(userDocument: DocumentNode, augmentedDocument: DocumentNode): DocumentNode {
+    const enricherContext = new EnricherContext(userDocument, augmentedDocument);
+    const enrichers: Enricher[] = [];
```

Comment:
discussion: I'm unsure about the `Enricher` name here. To me an enricher should only add things to an original, not modify it like in this case. 
However as you pointed out, there are really 2 behaviours here: adding types (really enriching) and modifying definitions, so probably one single name will never accurately reflect this.
What are your thoughts about something like this? 
```js
const transformer = new Transformer(userDocument); // when to specify the referenceDocument and document that gets modified are up for debate
transformer.registerEnricher(authorizationDefinitionEnricher)
transformer.registerEditor(authorizationUsageEnricher)
transformer.registerStatic(getStaticAuthorizationDefinitions())
return transformer.transform(augmentedDocument);
```
Even though `registerEnricher` and `registerEditor` might have the same implementation, it's clearer that they have different side effects and acommodates for different behaviours in the future if needed. Also then we'll contain the document transforming logic and the context inside the Transformer and apply it how we want to the transformer functions.

Reply:
The problem with Enricher and Editor being two separate entity is that even if now one enricher add new definitions and the other mutates the current definitions, this is only the current behavior of the authorizations enrichers while for others this can be done in a single pass. Considering that for every pass the whole Augmented Document is enumerated, having two separated entities (Enrichers and Editor) will lead to enumerating the Augmented Document twice, even when can be potentially achievable in a single pass. Of course, this is an initial design of this architecture component that can and very likely will be revisited during the refactoring of the augmented schema that will unlock different options!